cpp/src/copying/copy.cu

```diff
@@ -22,30 +22,74 @@ namespace cudf {
 namespace experimental {
 namespace detail {
 
-std::unique_ptr<column> copy_if_else( column_view const& lhs, column_view const& rhs, column_view const& boolean_mask,
+// column/column, scalar/column
+template <typename Left, typename Right>
+std::unique_ptr<column> copy_if_else( Left const& lhs, Right const& rhs, 
```

Comment:
This function is the point at which `lhs/rhs` should be converted into iterators. `make_pair_iterator` should work for both column/scalars. 

This would significantly simplify and eliminate many of the underlying code paths.